warehouse/accounts/auth_policy.py

```diff
@@ -10,15 +10,110 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import struct
+
+from pymacaroons import Macaroon, Verifier
+from pymacaroons.caveat_delegates import ThirdPartyCaveatVerifierDelegate
+from pymacaroons.exceptions import MacaroonException
+
 from pyramid.authentication import (
     BasicAuthAuthenticationPolicy as _BasicAuthAuthenticationPolicy,
+    CallbackAuthenticationPolicy as _CallbackAuthenticationPolicy,
     SessionAuthenticationPolicy as _SessionAuthenticationPolicy,
 )
 
 from warehouse.accounts.interfaces import IUserService
 from warehouse.cache.http import add_vary_callback
 
 
+class AccountTokenAuthenticationPolicy(_CallbackAuthenticationPolicy):
+    def __init__(self, authenticate):
+        self._authenticate = authenticate
+        self.callback = self._auth_callback
+
+        self._routes_allowed = ["forklift.legacy.file_upload"]
+
+    def unauthenticated_userid(self, request):
+        if request.matched_route.name not in self._routes_allowed:
+            return None
+
+        account_token = request.params.get("account_token")
+
+        if account_token is None:
+            return None
+
+        try:
+            macaroon = Macaroon.deserialize(account_token)
+
+            # First, check identifier and location
+            if macaroon.identifier != request.registry.settings["account_token.id"]:
+                return None
```

Comment:
We can definitely have a per-macaroon secret instead of one for all of Warehouse; that's a good best-practice.

It's a minor point, but in this implementation it is possible for a user to generate multiple macaroons and revoke a particular macaroon by deleting its row from the database. After we verify the macaroon, we use the `accounts_token` table to map the macaroon's id to a username, so if we remove that row the macaroon will no longer be usable.

Reply:
Yea, I noticed that later on, and then just forgot to update this review :)